sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala

```diff
@@ -1272,7 +1272,8 @@ class AstBuilder(conf: SQLConf) extends SqlBaseBaseVisitor[AnyRef] with Logging
    * Create a [[Cast]] expression.
    */
   override def visitCast(ctx: CastContext): Expression = withOrigin(ctx) {
-    Cast(expression(ctx.expression), visitSparkDataType(ctx.dataType))
+    Cast(expression(ctx.expression()), visitSparkDataType(ctx.dataType),
+      format = Option(ctx.STRING()).map(string))
```

Comment:
Actually, does it make sense to have another class with Cast as parent for this code specifically? Looks it's only one place that the format is needed.

Reply:
and the codes will be much less invasive.

Reply:
Yes, it could be implemented in such way. Just need to extract common code to a separate class or trait because _you cannot inherit from case class_, or make `Cast` as regular class (need to fix all matches).

Reply:
I'm fine with making that change and have a new `FormattedCast` but had some concerns here: https://github.com/apache/spark/pull/24742#discussion_r289281432

Reply:
Ah, right. inheritance of case class to case class is prohibited. but can't we inherit case class at a regular class?

Reply:
Or ... can we make a dummy case that uses `RuntimeReplaceable` only when the format is specified?

Reply:
Just wanted to make this change less invasive and wondering if this is the only way.

Reply:
I think it would make sense to pattern match by the base class, and make access to only used fields via base class/trait methods. This is even recommended by style guide (the last item): https://github.com/databricks/scala-style-guide#pattern-matching

Reply:
I reverted `Cast` to its original signature and introduced `CastBase` sealed trait and `FormattedCast` expression in the last commit.

@MaxGekk I also added an extractor to `CastBase` as handling complex pattern matching like `case e @ WindowExpression(Cast(Literal(0L, _), _, _, _), _)` becomes `case e @ WindowExpression(CastBase(Literal(0L, _), _), _)` with the new extractor and it is much easier than matching on type and using fields as the style guide suggests. Please let me know if you disagree.

@HyukjinKwon I'm very sorry but I don't see how `RuntimeReplaceable` should come into play. Could you please elaborate on that?

Reply:
I was thinking about using `RuntimeReplaceable`, for instance,

```scala
class CastWithFormat(... dataType: DataType, child: Expression) extends RuntimeReplaceable
  def this(c: Expression, format: ...)
    if dateType is DateType:
        this(c, dataType, DateFormatClass(c, format))
    if dateType is TimestampType:
        ...
}
```

Reply:
would it be possible?

Reply:
Ok, I see it now. Thanks!